sonyc_kalman/data.py

```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
     return X, mask
 
 
+def series_splitter(invalid_mask, test_ratio=0.25, mode='random_holes', hole_param=[20, 10]):
+    '''
+    Given a `invalid_mask` of length n, generate masks for the train split and test split according to `test_ratio`.
+
+    Params:
+    -------
+        invalid_mask: list or np.array
+            an array of 1's and 0's. 0 being valid, 1 being invalid
+        test_ratio: float (0, 1)
+            the fraction of the data to be split out as test
+        mode: str
+            should be one of 'random_holes', 'r', 'chronological', or 'c'.
+            'random_holes' - punch_out randomly (Guasssian according to `hole_param`) sized holes in the series and 
+            set aside as test.
+            'chronological' - just use the begining as train and the end as test, modeling real world delopyment.
+            hole_param in chronological mode is ignored.
+        hole_param: list [mean, std]
+            used for the 'random_holes' mode.
+            
+    Returns:
+    --------
+        train_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 means selected for the training split
+        test_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 means selected for the test split
+    '''
+    if mode == 'chronological' or mode == 'c':
+        chrono_mode = True
+    elif mode == 'random_holes' or mode == 'r':
+        chrono_mode = False
+    else:
+        raise ValueError("type must be one of 'random_holes', 'r', 'chronological', or 'c'.")
+
+    invalid_mask = np.array(invalid_mask, dtype=np.bool)
+    valid_mask = 1 - invalid_mask
+    valid_mask = valid_mask.astype(np.bool)
+    n_total_valid = sum(valid_mask)
+    n_test = int(round(test_ratio * n_total_valid))
+    n_train = n_total_valid - n_test
+    
+    train_mask = valid_mask.copy()
+    test_mask = valid_mask.copy()
+    if chrono_mode:
+        # first split the data as if all the valid frames are contiguous
+        continuous_split_test = np.array([0] * n_train + [1] * n_test, dtype=int)
+        continuous_split_train = 1 - continuous_split_test
+        # then use the valid_mask to put these contiguous masks into the right places
+        test_mask[valid_mask]  = continuous_split_test
+        train_mask[valid_mask] = continuous_split_train
+
+    else:
+        # 'random_holes' mode
+        ## first generate an array that holes the sizes of the holes, with length n_holes
+        hole_sizes = []
+        while sum(hole_sizes) < n_test:
+            rand_size = int(round(np.random.normal(hole_param[0], hole_param[1])))
```

Comment:
Also it might be good to also have deterministic holes so that we can evaluate performance as a function of hole size. Though as I type this I realize we can always just set the "scale" parameter to be 0.